runtime/compiler/ilgen/Walker.cpp

```diff
@@ -5061,13 +5061,54 @@ TR_J9ByteCodeIlGenerator::loadAuto(TR::DataType type, int32_t slot, bool isAdjun
    push(load);
    }
 
+static bool isFieldResovled(TR::Compilation *comp, TR_ResolvedJ9Method * owningMethod, int32_t cpIndex, bool isStore)
+   {
+   uint32_t offset = 0;
+   TR::DataType type = TR::NoType;
+   bool isVolatile = true, isFinal = false, isPrivate = false, isUnresolvedInCP;
+   bool bResolved = owningMethod->fieldAttributes(comp, cpIndex, &offset, &type, &isVolatile, &isFinal,
+                                    &isPrivate, isStore, &isUnresolvedInCP, true /* needsAOTValidation */);
+   return bResolved;
+   }
+
 void
 TR_J9ByteCodeIlGenerator::loadInstance(int32_t cpIndex)
    {
    if (_generateReadBarriersForFieldWatch && comp()->compileRelocatableCode())
       comp()->failCompilation<J9::AOTNoSupportForAOTFailure>("NO support for AOT in field watch");
-   TR::SymbolReference * symRef = symRefTab()->findOrCreateShadowSymbol(_methodSymbol, cpIndex, false);
-   loadInstance(symRef);
+
+   TR_ResolvedJ9Method * owningMethod = static_cast<TR_ResolvedJ9Method*>(_methodSymbol->getResolvedMethod());
+
+   if (TR::Compiler->om.areValueTypesEnabled() &&
+       owningMethod->isFieldQType(cpIndex))
+      {
+      if (isFieldResovled(comp(), owningMethod, cpIndex, false))
+         {
+         if (owningMethod->isFieldFlattened(comp(), cpIndex, _methodSymbol->isStatic()))
+            {
+            // TODO choose the runtime helper or not. Maybe based on an option?
```

Comment:
I think this does need to be made real to facilitate debugging if test failures occur - the transform is optional. Would it make more sense to always generate the helpers and then have an ilgen opts pass to do the lowering? That way we could use a standard performTransformation guard to guard the transform to allow for easier debugging?

If that makes things too complicated then at least some kind of JIT option is necessary.

Reply:
As per discussion with @Leonardo2718 last week, this part of code is likely moved to the runtime helper change that @Leonardo2718 is implementing. That's why I haven't updated it yet. The plan for now is to switch to the runtime flattened field helper by using an environment variable. My understanding is something like below: 

```c++
         if (owningMethod->isFieldFlattened(comp(), cpIndex, _methodSymbol->isStatic()))
            {
            static char * enableRuntimeFlattenFieldPutfieldHelper = feGetEnv("TR_EnableRuntimeFlattenFieldPutfieldHelper");
            if (enableRuntimeFlattenFieldPutfieldHelper)
               {
               // Runtime helper 
               }
            else
               {
               storeInstanceFlattenedField(cpIndex);
               }
            }
         else
            {
            TR::SymbolReference * symRef = symRefTab()->findOrCreateShadowSymbol(_methodSymbol, cpIndex, false);
            storeInstance(symRef);
            }
         }
```

Reply:
for debugging a performTransformation would be preferable to an envvar since the use of the fully flattened version is an optimization over the helper version right? eg we should degrade to the simpler one to help debug problems potentially? Thoughts? FYI @Leonardo2718 and @hzongaro

Reply:
That is my understanding too: by default we go down the transformation path, if anything wrong happens, we use the environment variable to choose the runtime helpers for debugging. I could be wrong tho. I'll let @Leonardo2718 help confirm.

Reply:
>That is my understanding too: by default we go down the transformation path, if anything wrong happens, we use the environment variable to choose the runtime helpers for debugging

I missunderstood the `performTransformation` macro and please ignore my comments on the above.

Reply:
After some discussion with @Leonardo2718 @hzongaro, JIT option is used for now to choose between runtime helper or sym ref generation during the compilation time (https://github.com/eclipse/omr/pull/5493). In the future, performTransformation could be considered. Moving the sym ref generation to Optimization phase is more complicated.